signature/src/balrog.rs

```diff
@@ -0,0 +1,363 @@
+        BalrogError::X509(X509Error::from(e))
+    }
+}
+
+impl From<PEMError> for BalrogError {
+    fn from(e: PEMError) -> Self {
+        BalrogError::PemDecodeError(e.to_string())
+    }
+}
+
+pub fn parse_pem_chain(input: &[u8]) -> Result<Vec<Pem>, PEMError> {
+    Pem::iter_from_buffer(input).collect::<Result<Vec<Pem>, PEMError>>()
+}
+
+/* A one-and-done function that parses a signature chain and validates a content signature. */
+pub fn parse_and_verify(
+    x5u: &[u8],
+    input: &[u8],
+    signature: &str,
+    current_time: i64,
+    root_hash: &str,
+    leaf_subject: &str,
+) -> Result<(), BalrogError> {
+    /* Parse the certificate chain. */
+    let pem_chain = parse_pem_chain(x5u)?;
+    let balrog = Balrog::new(&pem_chain)?;
+
+    /* Verify things. */
+    let _ = balrog.verify_chain(current_time, root_hash)?;
+    let _ = balrog.verify_leaf_hostname(leaf_subject)?;
+    let _ = balrog.verify_content_signature(input, signature)?;
+
+    /* Success */
+    Ok(())
+}
+
+impl<'a> Balrog<'_> {
+    pub fn new(list: &'a Vec<Pem>) -> Result<Balrog<'a>, BalrogError> {
+        if list.is_empty() {
+            return Err(BalrogError::CertificateNotFound);
+        }
+
+        let mut parsed: Vec<X509Certificate<'a>> = Vec::new();
+        let mut hash: Vec<u8> = Vec::new();
+        for (i, pem) in list.iter().enumerate() {
+            if pem.label != "CERTIFICATE" {
+                return Err(BalrogError::from(X509Error::InvalidCertificate));
+            }
+
+            let x509 = pem.parse_x509()?;
+            parsed.push(x509);
+
+            /* For the final certificate, calculate its SHA256 hash. */
+            /* TODO: Technically, we can get more bytes in the PEM beyond the end
+             * of the ASN.1 certificate sequence, and this function will include
+             * them in the hash. This feels like an unclear edge case though.
+             */
+            if i + 1 == list.len() {
+                let digest = digest::digest(&digest::SHA256, pem.contents.as_slice());
+                hash.extend_from_slice(digest.as_ref());
+            }
+        }
+
+        Ok(Balrog {
+            chain: parsed,
+            root_hash: hash,
+        })
+    }
+
+    /* Ensure that the certificate chain is valid. */
+    pub fn verify_chain(&self, current_time: i64, root_hash: &str) -> Result<(), BalrogError> {
+        if self.chain.is_empty() {
+            return Err(BalrogError::CertificateNotFound);
+        }
+
+        /* For each certificate N in the chain, check that:
+         *  1. The issuer name in cert N matches the subject name of cert N+1.
+         *  2. The validity time of cert N contains the current time.
+         *  3. The cert N+1 contains a key usage extension permitting CA.
+         *  4. Extract the public key from cert N+1
+         *  5. Call X509Certificate::verify_signature with the key from step 4.
+         *
+         * For the last certificate in the chain (the root), repeat the same
```

Comment:
Question: is it possible the chain will have only one certificate and is that an issue?

Reply:
It would be poor security practice for the SRE team to use the root certificate for code signing directly, but in terms of whether this breaks our trust in the update: it's fine.

Reply:
To elaborate, the root certificate is the considered to be very important. Anyone who can impersonate the root can effectively do anything, so a leak of the signing keys for this certificate would be catastrophic. Therefore, it's good practice to keep the signing keys for the root offline and in secure storage to minimize the surface area in which to attack them. Instead the leaf and intermediate certificate authorities are used to handle the day-to-day operational needs for code signing.

Reply:
Got it. Thanks for the explanation!